src/Test/Diagnostics/TestingOnly_WaitingService.cs

```diff
@@ -44,15 +56,43 @@ public void WaitForAsyncOperations(string featureName, bool waitForWorkspaceFirs
             GC.KeepAlive(featureWaiter);
         }
 
-        public void WaitForAllAsyncOperations(params string[] featureNames)
+        public ImmutableArray<KeyValuePair<string, TimeSpan>> WaitForAllAsyncOperations(params string[] featureNames)
         {
+            var longRunningTokens = new Dictionary<string, TimeSpan>();
+            var stopwatch = Stopwatch.StartNew();
             var task = _provider.WaitAllAsync(
                 featureNames,
 #pragma warning disable VSTHRD001 // Avoid legacy thread switching APIs
-                eventProcessingAction: () => Dispatcher.CurrentDispatcher.Invoke(() => { }, DispatcherPriority.ApplicationIdle));
+                eventProcessingAction: () => Dispatcher.CurrentDispatcher.Invoke(
+                    () =>
+                    {
+                        var foregroundNotificationService = _foregroundNotificationService as ForegroundNotificationService;
+                        foregroundNotificationService?.ReleaseCancelledItems();
```

Comment:
looks like ReleaseCancelledItems has a race unless this is called from UI thread since things can be enqueued after release cancelled items are called? looks like foreground notification should shutdown (don't enqueue new work and dispose asynctoken right away) after ReleaseCancelledItems is called

Reply:
actually, there might still race even if it is called from UI thread.

Reply:
ReleaseCancelledItems does only manipulates the queue from inside a lock:

https://github.com/dotnet/roslyn/blob/de57f6daeee86908cc9dcd0386d2f66d60e8cee1/src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs#L404